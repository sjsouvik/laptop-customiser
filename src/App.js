import { useState, useEffect } from "react";
import { Container } from "./common/components";
import { Header } from "./common/components";
import { Description } from "./common/components";
import { Price } from "./common/components";
import { segregateBasedOnType } from "./common/helper/utils";
import { getDefaultPrice } from "./service";

const App = () => {
  const [loading, setLoading] = useState(false);

  const [configList, setConfigList] = useState({});
  const [selectedConfigs, setSelectedConfigs] = useState({});

  const [error, setError] = useState(null);
  const [priceLoading, setPriceLoading] = useState(false);
  const [price, setPrice] = useState(0);
  const [priceError, setPriceError] = useState(null);

  useEffect(() => {
    const getMacComponents = () => {
      return fetch(`http://localhost:3004/components`)
        .then((response) => response.json())
        .then((data) => {
          const configs = data.reduce((acc, config) => {
            if (config.type === "proe") {
              acc = segregateBasedOnType(config, acc, "Processor");
            } else if (config.type === "mem") {
              acc = segregateBasedOnType(config, acc, "Memory");
            } else if (config.type === "gfx") {
              acc = segregateBasedOnType(config, acc, "Graphics");
            } else if (config.type === "stg") {
              acc = segregateBasedOnType(config, acc, "Storage");
            }

            return acc;
          }, {});

          setConfigList(configs);

          setSelectedConfigs({
            Processor: configs["Processor"][0],
            Memory: configs["Memory"][0],
            Graphics: configs["Graphics"][0],
            Storage: configs["Storage"][0],
          });
          setLoading(false);
        })
        .catch(() => {
          setError(new Error("could not fetch the customisable components"));
          setLoading(false);
        });
    };
    setLoading(true);
    getMacComponents();
    setPriceLoading(true);
    getDefaultPrice()
      .then((data) => {
        setPrice(data.value);
      })
      .catch((error) => {
        setPriceError(error);
      })
      .finally(() => {
        setPriceLoading(false);
      });
  }, []);

  const setSelectedVariant = ({ title, serialNo, variant, addOnPrice }) => {
    setSelectedConfigs((configs) => ({
      ...configs,
      [title]: { serialNo, variant, addOnPrice },
    }));
  };

  return (
    <>
      <Header />
      <main>
        <div className="main__container">
          <div className="main__content">
            <section>
              <img
                className="macbook-img"
                alt="macbook pro"
                src="https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/mbp16touch-space-select-201911?wid=1808&hei=1686&fmt=jpeg&qlt=80&.v=1572825197207"
              />
            </section>
            <section className="configuration">
              {loading && <h1>loading...</h1>}
              {error && <h1>Something went wrong. Please try again later</h1>}

              {!(loading || error) && (
                <>
                  <Description selectedConfigs={selectedConfigs} />
                  {Object.entries(configList).map(
                    ([configTitle, configOptions]) => (
                      <Container
                        key={configTitle}
                        title={configTitle}
                        list={configOptions}
                        selectedConfigs={selectedConfigs}
                        setSelectedVariant={setSelectedVariant}
                      />
                    )
                  )}
                </>
              )}
            </section>
          </div>
        </div>
        <Price
          price={price}
          priceLoading={priceLoading}
          priceError={priceError}
          selectedConfigs={selectedConfigs}
        />
      </main>
    </>
  );
};

export default App;
