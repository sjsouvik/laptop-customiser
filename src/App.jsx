import { useState, useEffect } from "react";
import { Header, Description, Container, Price } from "./common/components";
import { segregateBasedOnType } from "./common/helper/utils";
import { getDefaultPrice } from "./service";

const App = () => {
  const [loading, setLoading] = useState(false);

  const [configList, setConfigList] = useState({});
  const [selectedConfigs, setSelectedConfigs] = useState({});

  const [error, setError] = useState(null);
  const [priceLoading, setPriceLoading] = useState(false);
  const [price, setPrice] = useState(0);
  const [priceError, setPriceError] = useState(null);

  useEffect(() => {
    const getMacComponents = () => {
      return fetch(`/components`)
        .then((response) => response.json())
        .then((data) => {
          const configs = data.reduce((formattedData, config) => {
            if (config.type === "proe") {
              return segregateBasedOnType(config, formattedData, "Processor");
            } else if (config.type === "mem") {
              return segregateBasedOnType(config, formattedData, "Memory");
            } else if (config.type === "gfx") {
              return segregateBasedOnType(config, formattedData, "Graphics");
            } else if (config.type === "stg") {
              return segregateBasedOnType(config, formattedData, "Storage");
            }
          }, {});

          setConfigList(configs);

          setSelectedConfigs({
            Processor: configs["Processor"][0],
            Memory: configs["Memory"][0],
            Graphics: configs["Graphics"][0],
            Storage: configs["Storage"][0],
          });
          setLoading(false);
        })
        .catch(() => {
          setError(new Error("could not fetch the customisable components"));
          setLoading(false);
        });
    };

    setLoading(true);
    getMacComponents();
    setPriceLoading(true);

    getDefaultPrice()
      .then((data) => {
        setPrice(data.value);
      })
      .catch((error) => {
        setPriceError(error);
      })
      .finally(() => {
        setPriceLoading(false);
      });
  }, []);

  const setSelectedVariant = ({ title, serialNo, variant, addOnPrice }) => {
    setSelectedConfigs((configs) => ({
      ...configs,
      [title]: { serialNo, variant, addOnPrice },
    }));
  };

  return (
    <>
      <Header />
      <main>
        <div className="main__container">
          <div className="main__content">
            <section>
              <img
                className="macbook-img"
                alt="macbook pro"
                src="https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/mbp16touch-space-select-201911?wid=1808&hei=1686&fmt=jpeg&qlt=80&.v=1572825197207"
              />
            </section>
            <section className="configuration">
              {loading && <h1>loading...</h1>}
              {error && <h1>Something went wrong. Please try again later</h1>}

              {!(loading || error) && (
                <>
                  <Description selectedConfigs={selectedConfigs} />
                  {Object.entries(configList).map(
                    ([configTitle, configOptions]) => (
                      <Container
                        key={configTitle}
                        title={configTitle}
                        list={configOptions}
                        selectedConfigs={selectedConfigs}
                        setSelectedVariant={setSelectedVariant}
                      />
                    )
                  )}
                </>
              )}
            </section>
          </div>
        </div>
        <Price
          price={price}
          priceLoading={priceLoading}
          priceError={priceError}
          selectedConfigs={selectedConfigs}
        />
      </main>
    </>
  );
};

export default App;
