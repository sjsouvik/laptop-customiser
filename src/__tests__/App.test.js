import { fireEvent, render, waitFor, screen } from "@testing-library/react";
import App from "../App";
import data from "../../server/db.json";

const mockFetchCall = jest.spyOn(global, "fetch");

describe("Laptop customiser", () => {
  beforeEach(() => {
    mockFetchCall
      .mockResolvedValueOnce({
        json: () => Promise.resolve(data.components),
      })
      .mockResolvedValueOnce({
        json: () => Promise.resolve(data.price),
      });
  });

  it("should show the correct price when processor a is selected", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      fireEvent.click(getByTestId(`Processor_a`));
    });
    expect(getByTestId("Processor-details")).toHaveTextContent(
      "2.3GHz 8-core 9th-generation Intel Core processor, Turbo Boost up to 4.8GHz"
    );
    expect(getByTestId("total-price")).toHaveTextContent(`₹239900`);
  });

  it("should show the correct price when processor b is selected", async () => {
    const { getByTestId } = render(<App />);
    await waitFor(() => {
      fireEvent.click(getByTestId(`Processor_b`));
    });
    expect(getByTestId("Processor-details")).toHaveTextContent(
      "2.4GHz 8-core 9th-generation Intel Core processor, Turbo Boost up to 5.0GHz"
    );
    expect(getByTestId("total-price")).toHaveTextContent(`₹259900`);
  });

  it("renders correctly", async () => {
    const { container } = render(<App />);
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});

describe("Tests error scenarios", () => {
  it("should show error if getDefaultPrice service fails", async () => {
    mockFetchCall.mockRejectedValue().mockRejectedValue();

    const { getByText } = render(<App />);

    await waitFor(() => {
      expect(
        getByText(`Something went wrong. Please try again later`)
      ).toBeTruthy();
    });
  });
});
