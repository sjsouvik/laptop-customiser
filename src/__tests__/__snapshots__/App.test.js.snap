// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Laptop customiser renders correctly 1`] = `
<div>
  <header>
    <div
      class="header__content"
    >
      <a
        class="header__link"
        href="https://www.apple.com/in/macbook-pro"
        rel="noreferrer"
        target="_blank"
      >
        <strong>
          MacBook Pro
        </strong>
      </a>
    </div>
  </header>
  <main>
    <div
      class="main__container"
    >
      <div
        class="main__content"
      >
        <section>
          <img
            alt="macbook pro"
            class="macbook-img"
            src="https://store.storeimages.cdn-apple.com/4668/as-images.apple.com/is/mbp16touch-space-select-201911?wid=1808&hei=1686&fmt=jpeg&qlt=80&.v=1572825197207"
          />
        </section>
        <section
          class="configuration"
        >
          <h1>
            loading...
          </h1>
        </section>
      </div>
    </div>
    <div
      class="price container"
    >
      <div
        class="price__content"
      >
        <h1
          class="price__value"
        >
          Total:
           loading...
        </h1>
      </div>
    </div>
  </main>
</div>
`;
